lighthouse-core/audits/byte-efficiency/cache-headers.js

```diff
@@ -0,0 +1,229 @@
+
+class CacheHeaders extends ByteEfficiencyAudit {
+  /**
+   * @return {number}
+   */
+  static get WASTED_BYTES_DISCOUNT_MULTIPLIER() {
+    return WASTED_BYTES_DISCOUNT_MULTIPLIER;
+  }
+
+  /**
+   * @return {!AuditMeta}
+   */
+  static get meta() {
+    return {
+      category: 'Caching',
+      name: 'cache-headers',
+      informative: true,
+      helpText:
+        'A well-defined cache policy can speed up repeat visits to your page. ' +
+        '[Learn more](https://developers.google.com/speed/docs/insights/LeverageBrowserCaching).',
+      description: 'Leverage browser caching',
+      requiredArtifacts: ['devtoolsLogs'],
+    };
+  }
+
+  /**
+   * Converts a time in seconds into a duration string, i.e. `1d 2h 13m 52s`
+   * @param {number} maxAgeInSeconds
+   * @return {string}
+   */
+  static toDurationDisplay(maxAgeInSeconds) {
+    if (maxAgeInSeconds === 0) {
+      return 'None';
+    }
+
+    const parts = [];
+    const unitLabels = [
+      ['d', SECONDS_IN_DAY],
+      ['h', SECONDS_IN_HOUR],
+      ['m', SECONDS_IN_MINUTE],
+      ['s', 1],
+    ];
+
+    for (const [label, unit] of unitLabels) {
+      const numberOfUnits = Math.floor(maxAgeInSeconds / unit);
+      if (numberOfUnits > 0) {
+        maxAgeInSeconds -= numberOfUnits * unit;
+        parts.push(`${numberOfUnits}\xa0${label}`);
+      }
+    }
+
+    return parts.join(' ');
+  }
+
+  /**
+   * Computes the percent likelihood that a return visit will be within the cache lifetime, based on
+   * Chrome UMA stats see the note above.
+   * @param {number} maxAgeInSeconds
+   * @return {number}
+   */
+  static getCacheHitLikelihood(maxAgeInSeconds) {
+    const maxAgeInHours = maxAgeInSeconds / 3600;
+    const upperDecileIndex = RESOURCE_AGE_IN_HOURS_DECILES.findIndex(
+      decile => decile >= maxAgeInHours
+    );
+    if (upperDecileIndex === 11) return 1;
+    if (upperDecileIndex === 0) return 0;
+
+    const upperDecile = RESOURCE_AGE_IN_HOURS_DECILES[upperDecileIndex];
+    const lowerDecile = RESOURCE_AGE_IN_HOURS_DECILES[upperDecileIndex - 1];
+
+    const tenthsPlace = upperDecileIndex;
+    // approximate the position between deciles as linear
+    const hundredthsPlace = 10 * (maxAgeInHours - lowerDecile) / (upperDecile - lowerDecile);
+    return tenthsPlace / 10 + hundredthsPlace / 100;
+  }
+
+  /**
+   * Computes the user-specified cache lifetime, 0 if explicit no-cache policy is in effect, and null if not
+   * user-specified. See https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html.
+   *
+   * @param {!Map<string,string>} headers
+   * @param {!Object} cacheControl Follows the potential settings of cache-control, see https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control
+   * @return {?number}
+   */
+  static computeCacheLifetimeInSeconds(headers, cacheControl) {
+    // Pragma controls caching if cache-control is not set, see https://tools.ietf.org/html/rfc7234#section-5.4
```

Comment:
sure, done. seemed simpler to me to get the crazy old case out of the way first so that the rest of the function you know you're dealing with cache-control/expires precedence but I'm not the one having to read it from the outside :)